lib/Dialect/TensorExt/IR/TensorExtAttributes.cpp

```diff
@@ -0,0 +1,93 @@
+#include "lib/Dialect/TensorExt/IR/TensorExtAttributes.h"
+
+#include "llvm/include/llvm/ADT/STLExtras.h"   // from @llvm-project
+#include "mlir/include/mlir/IR/Diagnostics.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/MLIRContext.h"  // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"    // from @llvm-project
+
+namespace mlir {
+namespace heir {
+namespace tensor_ext {
+
+LogicalResult AlignmentAttr::verify(
+    ::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError,
+    mlir::detail::DenseArrayAttrImpl<long> in,
+    mlir::detail::DenseArrayAttrImpl<long> out,
+    mlir::detail::DenseArrayAttrImpl<long> padding,
+    mlir::detail::DenseArrayAttrImpl<long> insertedDims,
+    TypedAttr paddingValue) {
+  if (in.empty() || out.empty()) {
+    return emitError() << "in and out may not be empty arrays";
+  }
+
+  if (in.size() + insertedDims.size() != out.size()) {
+    return emitError()
+           << "in.size() + insertedDims.size() must equal out.size()";
+  }
+
+  if (!padding.empty() && padding.size() != out.size()) {
+    return emitError() << "padding.size() must equal out.size()";
+  }
+
+  for (auto val : padding.asArrayRef()) {
+    if (val < 0) {
+      return emitError() << "padding must be non-negative";
+    }
+  }
+
+  DenseSet<long> insertedDimsSet(insertedDims.asArrayRef().begin(),
+                                 insertedDims.asArrayRef().end());
+  if (insertedDimsSet.size() != insertedDims.size()) {
+    return emitError() << "insertedDims must all be unique";
+  }
```

Comment:
for the sake of verification, these probably should be within bounds as well? The code below that sets beforeReplication doesn't mind if they're out of range.

Reply:
Fixed in 4a99b02a